source/logging/logging.rst

```diff
@@ -0,0 +1,380 @@
+     - High-level information about normal driver behavior. Example:
+       ``MongoClient`` creation or close.
+     - ``info``
+
+   * - 7
+     - Debug
+     - Detailed information that may be helpful when debugging an application.
+       Example: A command starting.
+     - ``debug``
+
+   * - 8
+     - Trace
+     - Very fine-grained details related to logic flow. Example: entering and
+       exiting function bodies.
+     - ``trace``
+
+Note that the Emergency, Alert, and Critical levels have been intentionally left undefined. At the
+time of writing this specification, we do not expect any driver specifications to need to log at
+these levels, but we have included them in the list of permitted levels for consistency with Syslog.
+
+The levels above are defined in order from most to least severe. Not all logging frameworks will
+necessarily support all of these levels. If an equivalent level is not available, drivers MUST emit
+messages for that level at the closest, less severe level if one is available, or the closest more
+severe level otherwise.
+
+For example:
+
+- If an Informational level is not available and Debug is, Informational messages defined as \
+  Informational in a specification MUST be emitted at Debug level.
+- If a Trace level is not available, Trace messages MUST be omitted at Debug level.
+
+Structured versus Unstructured Logging
+--------------------------------------
+If structured logging is available in and idiomatic for the driver's language/ecosystem, the driver
+SHOULD produce structured log messages. Note that some ecosystems with structured logging support
+may also have support available to convert structured output to traditional unstructured messages
+for users who want it (for example, the
+`log feature <https://docs.rs/tracing/latest/tracing/#emitting-log-records>`_ in Rust's
+`tracing <https://docs.rs/tracing/latest/tracing/>`_ crate). If such support is available, drivers
+SHOULD utilize it to support both types of logging.
+
+Note that drivers implementing unstructured logging MUST still support some internal way to intercept
+the data contained in messages in a structured form. See the `unified test
+format specification <../unified-test-format/unified-test-format.rst#expectedLogMessage>`_ for details.
+
+Representing Errors in Log Messages
+-----------------------------------
+Drivers MAY represent errors in log messages in whatever format is idiomatic for their language and
+existing error types. For example, if a driver's error classes have existing ``toString()``
+implementations, those MAY be used. Alternatively, if a driver emits structured log messages, a
+structured format containing error data could be used. Any information which a driver reports via
+its error classes MUST be included in the log representations. 
+Note that if the driver includes full server responses in its errors these MUST be truncated in
+accordance with the max document length option.
+
+Performance Considerations
```

Comment:
Consider adding something like:

> Drivers SHOULD optimize JSON generation such that the complexity is O(N) where N = `<max document length>`, rather than N = `<actual document length>`